### PR TITLE
Include all Dockerfiles in the project and switch to GitHub Actions and Packages

### PR DESCRIPTION
**Pull request by anskarl**
### Docker images

   - Makefile for building the required docker images
   - `./services.sh start`: when images are not available will try to build the images locally
   - Alpine-based images for Apache druid and chaos-proxy
   - Openresty, Druid, Zookeeper and Chaos-proxy versions can be customized via arguments in Makefile
   - Github docker registry
   - Default Druid version is 0.16.0

### Switch to GitHub actions

 - similar to the current CircleCI workflow:
   * workflow is triggered on push or pull request
   * .ivy2 directory is cached (hashed base on the checksum of build.sbt)
   * coverage report is being archived
 - adds docker image publishing to github packages:
   * credentials are managed by github
   * no need to have separate user/space on Docker Hub
 - use of common .env file for both docker-compose and Makefile. The versions of external services (e.g., Druid) are controlled here
 - the workflow is portable, since forked repos will automatically publish images on their owner space
 - simplifications on Makefile (no need for latest tag, automatically load the contents of .env file)

### CODE REVIEW

**Comment by barend** [comment]
Hi @anskarl this is great stuff! I like GitHub actions (I actually have a half-done impl on my laptop, but I never got around to finishing it). I have made time to look into this and your other PR on Friday, together with @bjgbeelen. Sorry about the delayed response.

File: docker/druid/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+FROM $ARG_OPENJDK_IMAGE
+
+ARG ARG_DRUID_VERSION
+ARG ARG_ZOOKEEPER_VERSION
+
+ENV DRUID_VERSION=$ARG_DRUID_VERSION
+ENV ZOOKEEPER_VERSION=$ARG_ZOOKEEPER_VERSION
+ENV LANG C.UTF-8
+
+RUN apk add --no-cache bash perl python curl
```
Comment (barend) [inline docker/druid/Dockerfile]: This should install `python3` now, although that does require testing everything to see if it still works.

File: docker/Makefile [hunk 2]
```diff
@@ -0,0 +1,100 @@
+#
+# Parameters that can also be overridden from CLI
+#
+#   For example, to override the version of Druid to `0.15.1-incubating`:
+#
+#   make DRUID_VERSION=0.15.1-incubating image
+#
+
+ZOOKEEPER_VERSION ?= 3.4.14
```
Comment (barend) [inline docker/Makefile]: Latest is 3.5.6

File: docker/Makefile [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+DRUID_VERSION ?= 0.16.0-incubating
```
Comment (barend) [inline docker/Makefile]: Latest is 0.16.1-incubating

File: docker/Makefile [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+OPENRESTY_VERSION ?= 1.15.8.2-1
```
Comment (barend) [inline docker/Makefile]: Latest is 1.15.8.2-6

File: docker/Makefile [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+# ClusterF chaos proxy version is defined by the corresponding git commit hash id
+CHAOS_PROXY_VERSION ?= f935b60
```
Comment (barend) [inline docker/Makefile]: Latest is 39874835 (but that's a minor documentation update)

File: docker/chaos-proxy/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,24 @@
+ARG ARG_OPENJDK_IMAGE=openjdk:8-jre-alpine
+ARG ARG_BUILD_IMAGE=maven:3.6.1-jdk-8-alpine
```
Comment (barend) [inline docker/chaos-proxy/Dockerfile]: Latest is 3.6.3, although the jdk-8-alpine flavour no longer exists. It's either jdk-8-slim or ibmjava-8-alpine.

File: docker/openresty/Dockerfile [hunk 7]
```diff
@@ -1,4 +1,6 @@
-FROM openresty/openresty:1.15.8.2-1-alpine
+ARG ARG_OPENRESTY_VERSION=1.15.8.2-1
```
Comment (barend) [inline docker/openresty/Dockerfile]: Latest is 1.15.8.2-6-alpine

File: .env [hunk 8]
```diff
@@ -0,0 +1,5 @@
+ZOOKEEPER_VERSION=3.4.14
```
Comment (barend) [inline .env]: Latest is 3.5.6

File: .env [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+DRUID_VERSION=0.16.0-incubating
```
Comment (barend) [inline .env]: Latest is 0.16.1-incubating

File: .env [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+OPENRESTY_VERSION=1.15.8.2-1
```
Comment (barend) [inline .env]: Latest is 1.15.8.2-6

File: .env [hunk 11]
```diff
@@ -0,0 +1,5 @@
...
+# ClusterF chaos proxy version is defined by the corresponding git commit hash id
+CHAOS_PROXY_VERSION=f935b60
```
Comment (barend) [inline .env]: Latest is 39874835 (but that's only a tiny documentation change that doesn't impact this at all).

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+          restore-keys: |
+            ${{ runner.os }}-ivy2-
+      - name: Start dockerized services
+        run: ./services.sh start
+      - name: Wait Druid to become available
+        run: wget localhost:8082/status
+      - name: Build for Scala 2.11
+        run: sbt ++2.11.12 compile
+      - name: Build for Scala 2.12
+        run: sbt ++2.12.9 compile
```
Comment (barend) [inline .github/workflows/main.yml]: Latest is 2.12.10

**Comment by anskarl** [comment]
@barend thanks for the comments! I will make the version updates and remove the `.circleci` directory, possibly today or in the weekend.

**Comment by anskarl** [comment]
I've made the required changes (versions and removal of `.circleci` directory). 

I've also made some additional changes:
  
  - I found that I forgot to set the REPOSITORY_OWNER environment variable in the `docker-compose.yaml` file. Therefore, I've updated the file and did some minor adjustments to `services.sh` and `Makefile` for this environment variable.
  - I've updated the Dockerfiles to use `set -ex` at the beginning of each `RUN` expression
  - The build image of `clusterf-chaos-proxy` now uses `maven:3.6.3-jdk-8-slim` which is Debian-based, therefore the installation of required tools (git and openssh client) have been also adjusted.

With the upgrade of all image/tool versions and Dockerfiles, in practice we have an additional verification that the entire process of building, testing and publishing of docker images is automated and produces the expected results :)

**Comment by barend** [comment]
Looks good to me, thanks for updating. 

The Circle-CI fails because the configuration was removed, which is precisely the aim of this PR.

**Pull request merged**